### PR TITLE
[CASSANDRA-21354][trunk] Avoid serialization and deserialization for coordinator-local single partition data read

### PR DESCRIPTION
**Pull request by netudima**
Currently, when we execute a local read we fetch data from SSTables and Memtables using a merging iterator and write it to a byte buffer. Later when we combine a CQL response we deserialize the data back to iterate over them as a part of coordinator logic. So, we allocate rows and cells twice here, during the read from SSTables/Memtables and during the deserialization by coordinator logic if we read data locally (it is a typical scenario because usually drivers are sending requests to replicas).

The idea of optimization: if we do a single partition read of a small number of rows we can keep the data in memory and avoid double row objects allocation.

A system property is  used to limit number of rows we keep in memory in this scenario (to avoid too much pressure on GC due to extended lifetime for these objects and promoting them to an old generation). The property also allows to disable the logic in case of any issues.

We cannot get a number of rows in advance, so we read first N rows to memory and if we still have something then we serialize the remaining to a buffer and then concatenate iterators for the in-memory rows + deserialized one when we need to iterate over the result

patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-21354

### CODE REVIEW

File: src/java/org/apache/cassandra/db/ReadCommand.java [hunk 1]
```diff
@@ -436,14 +436,26 @@ public ReadCommand copyAsDigestQuery(Iterable<Replica> replicas)
     public abstract boolean isReversed();
 
     public ReadResponse createResponse(UnfilteredPartitionIterator iterator, RepairedDataInfo rdi)
+    {
+        return createResponse(iterator, rdi, false);
+    }
+    public ReadResponse createResponse(UnfilteredPartitionIterator iterator, RepairedDataInfo rdi, boolean localRead)
```
Comment (frankgh) [inline src/java/org/apache/cassandra/db/ReadCommand.java]: NIT line break?
```suggestion

    public ReadResponse createResponse(UnfilteredPartitionIterator iterator, RepairedDataInfo rdi, boolean localRead)
```

File: test/unit/org/apache/cassandra/db/ReadResponseTest.java [hunk 2]
```diff
@@ -172,6 +191,212 @@ public void makeDigestDoesntConsiderRepairedDataInfo()
...
+        // 3 rows, limit is 10 — all fit in memory with no overflow
+        testMultipleRows(10, 3);
+    }
+
+    @Test
+    public void inMemoryResponseWithZeroMaxRowsUsesOnlyOverflow()
+    {
+        // all rows go directly into the overflow buffer
+        testMultipleRows(0, 3);
+    }
```
Comment (frankgh) [inline test/unit/org/apache/cassandra/db/ReadResponseTest.java:820]: We are missing a test that has a RT marker between in memory rows and the overflow rows. I don't think we are handling this case correctly. We have a potential data resurrection issue here. Can we add this test?
```suggestion
    }

    @Test
    public void inMemoryResponseWithRangeTombstoneBetweenInMemoryAndOverflow()
    {
        // RT at [1, 2] sits between the in-memory rows (0) and the overflow rows (3-4)
        testWithTombstones(1, 5, 1, 2);
    }
```

The fix I suggest is in `org.apache.cassandra.db.ReadResponse.InMemoryDataResponse.LimitedUnfilteredRowIterator#computeNext` to ensure the RT open/close pairs are never split . We either keep them entirely in memory or in overflow:

```
            private boolean insideOpenMarker = false;

            @Override
            protected Unfiltered computeNext()
            {
                if (!wrapped.hasNext())
                    return endOfData();

                if (rowCount >= maxRows && !insideOpenMarker)
                {
                    hasOverflow = true;
                    return endOfData();
                }

                Unfiltered next = wrapped.next();
                if (next.isRangeTombstoneMarker())
                {
                    RangeTombstoneMarker marker = (RangeTombstoneMarker) next;
                    insideOpenMarker = marker.isOpen(isReverseOrder);
                }
                else
                    rowCount++;
                return next;
            }
```

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 3]
```diff
@@ -288,8 +325,8 @@ static abstract class DataResponse extends ReadResponse
         // TODO: can the digest be calculated over the raw bytes now?
         // The response, serialized in the current messaging version
         private final ByteBuffer data;
-        private final ByteBuffer repairedDataDigest;
-        private final boolean isRepairedDigestConclusive;
+        protected ByteBuffer repairedDataDigest;
+        protected boolean isRepairedDigestConclusive;
```
Comment (frankgh) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: can we keep these as `private final` and instead provide a static method to `build` an `InMemoryDataResponse` object? something like this:

```
        static InMemoryDataResponse build(UnfilteredPartitionIterator iter, ReadCommand command, RepairedDataInfo rdi, int maxRows)
        {
            ImmutableBTreePartition partition;
            ByteBuffer overflow;

            if (!iter.hasNext())
            {
                partition = null;
                overflow = null;
            }
            else
            {
                try (UnfilteredRowIterator rowIter = iter.next())
                {
                    LimitedUnfilteredRowIterator limitedIter = new LimitedUnfilteredRowIterator(rowIter, maxRows);
                    partition = ImmutableBTreePartition.create(limitedIter);
                    // Uses buildOverflow (row-iterator level) to match the UnfilteredRowIteratorSerializer
                    // deserializer used in makeIterator.
                    overflow = limitedIter.hasOverflow()
                               ? LocalDataResponse.buildOverflow(rowIter, command.columnFilter())
                               : null;
                }
            }

            // Capture digest after consuming the iterator so that any updates made by
            // RepairedDataInfo transformations are reflected in the digest.
            return new InMemoryDataResponse(partition, overflow, rdi.getDigest(), rdi.isConclusive());
        }

        private InMemoryDataResponse(ImmutableBTreePartition partition, ByteBuffer overflow,
                                     ByteBuffer repairedDataDigest, boolean isRepairedDigestConclusive)
        {
            super(null, repairedDataDigest, isRepairedDigestConclusive, MessagingService.current_version, DeserializationHelper.Flag.LOCAL);
            this.partition = partition;
            this.overflow = overflow;
        }
```

**Review by frankgh — CHANGES_REQUESTED** [review]
There is a potential issue with data resurrection with this patch. I have added a few comments. Please take a look

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 3 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: thanks, good idea, I was too focused on trying to keep the logic in a constructor and missed an option with a factory method.

File: test/unit/org/apache/cassandra/db/ReadResponseTest.java (hunk 2 above)
Comment (netudima) [reply test/unit/org/apache/cassandra/db/ReadResponseTest.java:820]: thank you for highlighting the issue!

Comment (netudima) [reply test/unit/org/apache/cassandra/db/ReadResponseTest.java:820]: I've applied a very similar change + extended test coverage for this logic

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 4]
```diff
@@ -49,10 +58,43 @@ public abstract class ReadResponse
...
+    static final int IN_MEMORY_MAX_ROWS = CassandraRelevantProperties.DATA_RESPONSE_IN_MEMORY_MAX_ROWS.getInt();
+    static final long IN_MEMORY_MAX_SIZE = CassandraRelevantProperties.DATA_RESPONSE_IN_MEMORY_MAX_SIZE.getSizeInBytes();
+    static final int IN_MEMORY_MAX_ROWS_CL_ONE = CassandraRelevantProperties.DATA_RESPONSE_IN_MEMORY_MAX_ROWS_CL_ONE.getInt();
+    static final long IN_MEMORY_MAX_SIZE_CL_ONE = CassandraRelevantProperties.DATA_RESPONSE_IN_MEMORY_MAX_SIZE_CL_ONE.getSizeInBytes();
+
     protected ReadResponse()
     {
     }
 
+    private static int maxRows(boolean localReplicaOnly)
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: This method name is ambiguous in context of the class and can be misinterpreted by fellow humans. Perhaps follow the naming of the related constants instead?

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 5]
```diff
@@ -49,10 +58,43 @@ public abstract class ReadResponse
...
+    {
+        return localReplicaOnly ? IN_MEMORY_MAX_ROWS_CL_ONE : IN_MEMORY_MAX_ROWS;
+    }
+
+    private static long maxSize(boolean localReplicaOnly)
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: Ditto.

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 6]
```diff
@@ -49,10 +58,43 @@ public abstract class ReadResponse
     // Serializer for single partition read response
     public static final IVersionedSerializer<ReadResponse> serializer = new Serializer();
 
+    // Per-request limits for the in-memory portion of a local single-partition read response. Unfiltered objects are
+    // kept as an in-memory partition until either enabled limit is reached; the remainder is serialized into an
+    // overflow buffer. Each limit is disabled independently by setting it to <= 0.
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: Is this part of the comment still accurate?

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java [hunk 7]
```diff
@@ -207,6 +207,11 @@ public enum CassandraRelevantProperties
     DATA_OUTPUT_STREAM_PLUS_TEMP_BUFFER_SIZE("cassandra.data_output_stream_plus_temp_buffer_size", "8192"),
     DATA_RESPONSE_BUFFER_INITIAL_SIZE_MAX("cassandra.data_response_buffer_initial_size_max", "4096"),
     DATA_RESPONSE_BUFFER_INITIAL_SIZE_MIN("cassandra.data_response_buffer_initial_size_min", "128"),
+    /** higher in-memory limits used for ONE/LOCAL_ONE reads, when local result is consumed directly without waiting for other replicas **/
+    DATA_RESPONSE_IN_MEMORY_MAX_ROWS("cassandra.data_response_in_memory_max_rows", "128"),
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:211]: With byte-size limits implemented, is there really value in doing both row + byte size? I would imagine that byte-size is strictly superior, with row-size being quite arbitrary in practice (think huge collections, for example).

Comment (iamaleksey) [reply src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:211]: It might be, I'm just struggling to see how. So this is a genuine question, and  *not* a request to change disguised as a question.

File: src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.metrics;
+
+import com.codahale.metrics.Counter;
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+public class ReadResponseMetrics
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java:24]: I would argue that:

1. These metrics belong in `TableMetrics`, and
2. It's currently not straightforward at all to calculate what the hit ratio is

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 9]
```diff
@@ -354,6 +404,214 @@ public boolean isDigestResponse()
...
+        {
+            if (!iter.hasNext())
+            {
+                // Empty response
+                return new InMemoryDataResponse(null, rdi.getDigest(), rdi.isConclusive());
+            }
+
+            ImmutableBTreePartition partition;
+            ByteBuffer serialized;
+            try (UnfilteredRowIterator rowIter = iter.next())
```
Comment (iamaleksey) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: I think you end up closing this iterator twice (first time eventually inside `UnfilteredPartitionIterators.Serializer#serialize()`, and it has visible side-effects (metrics)).

**Review by iamaleksey — CHANGES_REQUESTED** [review]
First pass of this new version.

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 6 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: thanks, yes, the comment is out of date

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java (hunk 7 above)
Comment (netudima) [reply src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:211]: My main worry is about the case when we have many small rows and the number of objects is high but total heap size is not that big, counting of objects is too complicated and introduces extra overheads, so as a trade-off I decided to preserve the row check... From runtime overhead point of view row limit is cheap (compared to heap size calculation which requires object graph traversal).

File: src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java (hunk 8 above)
Comment (netudima) [reply src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java:24]: 1. Yes, I was thinking about it but decided not to go in this way because we already have a lot of Table-level metrics and this one I suppose will be rarely used to introduce such overhead + it requires to add a lookup of CFS object to get metrics object. At the same time the overheads are small, so if you think it is beneficial to have per table granularity here, I do not see a  problem to add it. 
2. Yes, we have currently org.apache.cassandra.metrics.ClientRequestMetrics#localRequests which helps to estimate the number of local replica reads, but it does not differentiate the cases then the local replica was actually queried + the single partition case, so it makes sense to add an extra metric here to see a ratio. From overhead point of view it is better to have misses count to still have one metric update per request, while from usability point of view total probably is slightly easier to use..

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java (hunk 7 above)
Comment (iamaleksey) [reply src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:211]: Fair enough.

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 9 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: good catch, thank you!

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 4 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: addressed

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 5 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: addressed

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 6 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: addressed

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 9 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: addressed, notes:
1) it does not reproduce with the current logic because BaseIterator.close has a protection against double close, at the same time it is a kind of implementation details, which is not guaranteed, so I've adjusted the logic to avoid a potential issue
2) I have not used a wrapping iterator to avoid extra object allocation as well as to not increase the number of types invoked on a hot path (extra risks to get metamorphic calls)

File: src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java (hunk 8 above)
Comment (netudima) [reply src/java/org/apache/cassandra/metrics/ReadResponseMetrics.java:24]: a counter for cases when a local single-partition read response kept in memory is added, to be able to see the ratio

**Review by iamaleksey — APPROVED** [review]
+1 with the last two issues addressed.

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java [hunk 10]
```diff
@@ -207,6 +207,11 @@ public enum CassandraRelevantProperties
     DATA_OUTPUT_STREAM_PLUS_TEMP_BUFFER_SIZE("cassandra.data_output_stream_plus_temp_buffer_size", "8192"),
     DATA_RESPONSE_BUFFER_INITIAL_SIZE_MAX("cassandra.data_response_buffer_initial_size_max", "4096"),
     DATA_RESPONSE_BUFFER_INITIAL_SIZE_MIN("cassandra.data_response_buffer_initial_size_min", "128"),
+    /** higher in-memory limits used for ONE/LOCAL_ONE reads, when local result is consumed directly without waiting for other replicas **/
```
Comment (frankgh) [inline src/java/org/apache/cassandra/config/CassandraRelevantProperties.java]: I think it's worth a comment that these limits won't be enforced across a range tombstone span (for correctness reasons).

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 11]
```diff
@@ -354,6 +411,317 @@ public boolean isDigestResponse()
...
+                    insideOpenMarker = marker.isOpen(isReverseOrder);
+                }
+                else
+                {
+                    next = withoutShadowedData((Row) next);
+                }
+                unfilteredCount++;
+                // Compute the heap size only when the size limit is enabled to avoid overheads if it is not needed
+                if (inMemoryMaxHeapSize > 0)
+                    accumulatedHeapSize += unsharedHeapSize(next);
```
Comment (frankgh) [inline src/java/org/apache/cassandra/db/ReadResponse.java:579]: we accumulate first and then check later in line 558. I think there could be cases where we exceed the limits (and we could potentially have a multi MiB row). I think we should at least document this is intentional.

File: src/java/org/apache/cassandra/db/ReadResponse.java [hunk 12]
```diff
@@ -354,6 +411,317 @@ public boolean isDigestResponse()
...
+        }
+
+        @Override
+        protected ByteBuffer getSerializedData()
+        {
+            throw new UnsupportedOperationException("InMemoryDataResponse cannot be serialized over the network");
+        }
+    }
+
+    private static class SingletonUnfilteredPartitionIterator extends AbstractUnfilteredPartitionIterator
```
Comment (frankgh) [inline src/java/org/apache/cassandra/db/ReadResponse.java]: Can't we use `org.apache.cassandra.db.partitions.SingletonUnfilteredPartitionIterator` instead? Metadata is derived from `iter.metadata()`. and the implementation is almost 1:1 to this implementation.

**Review by frankgh — COMMENTED** [review]
I have 3 smaller comments. Otherwise looks good

**Comment by frankgh** [comment]
We need to call out this change in NEWS.txt also

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 11 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java:579]: to measure heap size we have to consume an element from the original iterator, we cannot put it back easily. We do not keep it for a long time and almost immediate pass to serialization in case of overflow. Also we cannot deserialize a multi MiB row partially, so in any case we have to keep it in memory for while.

Comment (frankgh) [reply src/java/org/apache/cassandra/db/ReadResponse.java:579]: I think we just need to call it out

File: src/java/org/apache/cassandra/db/ReadResponse.java (hunk 12 above)
Comment (netudima) [reply src/java/org/apache/cassandra/db/ReadResponse.java]: I was worrying about deriving of TableMetadata from iterator vs using the value from command (as deserialisation logic do) because it could introduce some discrepancies in digest as we had for columns but it seems there is no such issue for TableMetadata.., so I will double check it and remove the code duplication.

**Comment by netudima** [comment]
added and rebased on fresh trunk

**Review by frankgh — APPROVED** [review]
+1

**Comment by netudima** [comment]
Merged as https://github.com/apache/cassandra/commit/72affbc20669f1b610dd2c997fc1f6c642f017c5
